Add metadata update ceremony

MVrachev opened the pull request:
Add metadata update ceremony.

Closes #222

kairoaraujo reviewed (commented):
This initial review is my first round of the UX.

kairoaraujo reviewed (commented):
I noticed an expected behavior:

```
Do you want to change the root metadata? [y/n]: n
Skipping further root metadata changes.

...

Do you want to change the online key? [y/n]: n
Skipping further online key changes.
```

It generates a new payload with root metadata version 2 (including new expirations)

kairoaraujo commented on repository_service_tuf/cli/admin/metadata.py:
```diff
@@ -0,0 +1,498 @@
+    console.print(markdown.Markdown(EXPIRY_CHANGES_MSG), width=100)
+    expiry_table = table.Table()
+    expiry_table.add_column("Role", justify="center")
+    expiry_table.add_column("Expiration Policy", justify="center")
```
I was evaluating it. Let's handle the expiration not in the Metadata Update feature.
There is no reason for us to implement it during the metadata rotation.

We can handle it using the REST API  POST in `/api/v1/config` and specific feature in the CLI.

In the Metadata Update, we coudl:

```mermaid
flowchart TD
    root[Do you want to change the root metadata?] --> root_answer{Y/N}
    root_answer-->|Yes| root_expire[Do you want update root expiration?]
    root_answer-->|No| online_keys[Online keys...]
    root_expire --> root_expire_answer{Y/N}
    root_expire_answer -->|Yes| root_new_expire_days[Exprire of days from now: 365]
    root_expire_answer -->|No| root_keys[Root Keys ... ]


```

MVrachev replied:
Ahaa, so updating only root expiration and we can create another end point to update the automatic bump expiry.
Let's have a discussion about what we want to document in that issue.

MVrachev replied:
Now, when we won't update the roles using the online key we can change the wording from `expiration policy` to `expiration bump` as it will show clearly that the expiration will be actually bumped.

kairoaraujo replied:
I closed the old:
- https://github.com/repository-service-tuf/repository-service-tuf-worker/issues/287

MVrachev commented on repository_service_tuf/cli/admin/metadata.py:
```diff
@@ -0,0 +1,498 @@
+        console.print(markdown.Markdown(ROOT_KEY_ADDITIONS_MSG), width=100)
+        _keys_additions(current_root)
+
+        console.print("\nHere is the current content of root:")
```
I am just wondering if I change this to: `Here is the new content of root:` as you suggested how will then sound when I ask the user: `Do you want to change the root metadata?`
Is it clear that it's about the same root?

kairoaraujo replied:
Yes, it sounds clear to me.

MVrachev commented on repository_service_tuf/cli/admin/metadata.py:
```diff
@@ -0,0 +1,498 @@
+        expiry_bump = _get_positive_int_input(
+            msg, "Expiration policy days", default
+        )
+        current_root.save_expiry(role, expiry_bump)
```
If you have given me a new expiration policy for root what am I supposed to do with it if not changing the root expiration as well?
I am asking a little different question for root:
`Expiration policy for root - number of days to bump expiration`
as compared to the others:
`Expiration policy for root - number of days to automatically bump expiration`

I will make it clearer for root by adding `now`:
`Expiration policy for root - number of days to bump expiration now`

MVrachev commented on repository_service_tuf/cli/admin/metadata.py:
```diff
@@ -0,0 +1,498 @@
+online key),
+- change the root threshold value
+
+The result of this ceremony will be a new root metadata file.
```
Okay, if we are going with the naming I will rename the default payload to an even more clear name like: `metadata-update-payload.json`.

kairoaraujo commented on repository_service_tuf/cli/admin/metadata.py:
```diff
@@ -0,0 +1,520 @@
+        )
+    try:
+        curr_root: RootInfo = _get_curr_root(current_root_uri)
+    except StorageError:
```
Do you want to add here `json.decoder.JSONDecodeError`?
It should be clear to the user that the file is incompatible.

See:
```shell
❯ rstuf admin metadata update
File name or URL to the current root metadata: Makefile
```
or
```shell
rstuf admin metadata update --current-root-uri Makefile
```

will fail like it
```shell
Traceback (most recent call last):
  File "/Users/kdearaujo/.local/share/virtualenvs/repository-service-tuf-cli-fl9agVA1/lib/python3.10/site-packages/tuf/api/serialization/json.py", line 36, in deserialize
    json_dict = json.loads(raw_data.decode("utf-8"))
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/json/__init__.py", line 346, in loads
    return _default_decoder.decode(s)
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/json/decoder.py", line 337, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/json/decoder.py", line 355, in raw_decode
    raise JSONDecodeError("Expecting value", s, err.value) from None
json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/kdearaujo/.local/share/virtualenvs/repository-service-tuf-cli-fl9agVA1/bin/rstuf", line 8, in <module>
    sys.exit(rstuf())
  File "/Users/kdearaujo/.local/share/virtualenvs/repository-service-tuf-cli-fl9agVA1/lib/python3.10/site-packages/click/core.py", line 1130, in __call__
    return self.main(*args, **kwargs)
  File "/Users/kdearaujo/.local/share/virtualenvs/repository-service-tuf-cli-fl9agVA1/lib/python3.10/site-packages/rich_click/rich_group.py", line 21, in main
    rv = super().main(*args, standalone_mode=False, **kwargs)
  File "/Users/kdearaujo/.local/share/virtualenvs/repository-service-tuf-cli-fl9agVA1/lib/python3.10/site-packages/click/core.py", line 1055, in main
    rv = self.invoke(ctx)
  File "/Users/kdearaujo/.local/share/virtualenvs/repository-service-tuf-cli-fl9agVA1/lib/python3.10/site-packages/click/core.py", line 1657, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/kdearaujo/.local/share/virtualenvs/repository-service-tuf-cli-fl9agVA1/lib/python3.10/site-packages/click/core.py", line 1657, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/kdearaujo/.local/share/virtualenvs/repository-service-tuf-cli-fl9agVA1/lib/python3.10/site-packages/click/core.py", line 1657, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/kdearaujo/.local/share/virtualenvs/repository-service-tuf-cli-fl9agVA1/lib/python3.10/site-packages/click/core.py", line 1404, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/kdearaujo/.local/share/virtualenvs/repository-service-tuf-cli-fl9agVA1/lib/python3.10/site-packages/click/core.py", line 760, in invoke
    return __callback(*args, **kwargs)
  File "/Users/kdearaujo/.local/share/virtualenvs/repository-service-tuf-cli-fl9agVA1/lib/python3.10/site-packages/click/decorators.py", line 26, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "/Users/kdearaujo/devel/RSTUF/repository-service-tuf-cli/repository_service_tuf/cli/admin/metadata.py", line 488, in update
    curr_root: RootInfo = _get_curr_root(current_root_uri)
  File "/Users/kdearaujo/devel/RSTUF/repository-service-tuf-cli/repository_service_tuf/cli/admin/metadata.py", line 115, in _get_curr_root
    curr_root_md = Metadata.from_file(curr_root_uri)
  File "/Users/kdearaujo/.local/share/virtualenvs/repository-service-tuf-cli-fl9agVA1/lib/python3.10/site-packages/tuf/api/metadata.py", line 234, in from_file
    return cls.from_bytes(file_obj.read(), deserializer)
  File "/Users/kdearaujo/.local/share/virtualenvs/repository-service-tuf-cli-fl9agVA1/lib/python3.10/site-packages/tuf/api/metadata.py", line 264, in from_bytes
    return deserializer.deserialize(data)
  File "/Users/kdearaujo/.local/share/virtualenvs/repository-service-tuf-cli-fl9agVA1/lib/python3.10/site-packages/tuf/api/serialization/json.py", line 40, in deserialize
    raise DeserializationError("Failed to deserialize JSON") from e
tuf.api.serialization.DeserializationError: Failed to deserialize JSON
```

kairoaraujo commented on repository_service_tuf/cli/admin/metadata.py:
```diff
@@ -0,0 +1,498 @@
+@click.option(
+    "--current-root",
+    help="URL or local path to the current root.json file.",
+    required=True,
```
I gave you the wrong sequence:
- `INTRODUCTION`
- Ask the user the `current-root-uri` (if not using the parameter `--current-root-uri`)
- `CURRENT_ROOT_INFO`

kairoaraujo commented on repository_service_tuf/cli/admin/metadata.py:
```diff
@@ -0,0 +1,520 @@
+- extending root expiration
+- modification of the keys used for signing (no matter if root keys or the
+online key),
+- changing of the root threshold value
```
```suggestion
- changing Root threshold value
```
nit: Name of the metadata

kairoaraujo commented on repository_service_tuf/cli/admin/metadata.py:
```diff
@@ -0,0 +1,532 @@
+
+from repository_service_tuf.cli import console
+from repository_service_tuf.cli.admin import admin
+from repository_service_tuf.cli.admin.ceremony import _load_key, _save_payload
```
We could do a structure change here:
- move `_load_key` function to `repository_service_tuf.helpers.tuf` as public `load_key`. Other CLI features could use it (`ceremony`, `metadata update` and [`key public-info`](https://github.com/repository-service-tuf/repository-service-tuf-cli/issues/290)) 
- move `_save_payload` function to `repository_service_tuf.helpers.tuf` as public `save_payload`. Other CLI features could use it

kairoaraujo replied:
About the first bullet, I implemented it in the below PR. You can use it from there.
- https://github.com/repository-service-tuf/repository-service-tuf-cli/pull/293

kairoaraujo reviewed (changes requested):
We must add the basic documentation in `docs/source/guide/index.rst`

Also, the development design in `docs/diagrams/repository-service-tuf-cli-C3.puml`

kairoaraujo commented on repository_service_tuf/cli/admin/metadata.py:
```diff
@@ -0,0 +1,520 @@
+The metadata update allows:
+- extending root expiration
+- modification of the keys used for signing (no matter if root keys or the
+online key),
```
The information inside the parentheses still needs to be changed. Please mark it as resolved if you want to keep it as is.

MVrachev commented:
I need to only add a comment adding documentation in `docs/source/guide/index.rst` and `docs/diagrams/repository-service-tuf-cli-C3.puml` as suggested by @kairoaraujo.

MVrachev commented:
I added:
- upload option to directly upload to RSTUF API after the ceremony is finished
- documentation 
- moved three functions as helper functions, so they can be reused by the metadata update ceremony code


I see that Codecov complains about test code coverage and I am starting writing tests, but should I propose them here given how big the pr became?

MVrachev commented on tests/unit/cli/admin/test_ceremony.py:
```diff
@@ -385,6 +88,44 @@ def test_ceremony_start_default_values(
+        assert "Ceremony done. 🔐 🎉." in test_result.output
+        # passwords not shown in output
+        assert "strongPass" not in test_result.output
+
```
I don't think I touched that test.
I think GitHub shows it incorrectly.

MVrachev commented on tests/unit/cli/admin/test_ceremony.py:
```diff
@@ -385,6 +88,44 @@ def test_ceremony_start_default_values(
         # passwords not shown in output
         assert "strongPass" not in test_result.output
 
+    def test_ceremony_negative_expiry_and_try_again(
```
I don't think I touched that test.
I think GitHub shows it incorrectly.

lukpueh reviewed (commented):
Here's a first set of comments. I've only reviewed the workflow based on index.rst so far. This is very cool stuff! I'll comment more tomorrow! :)

lukpueh commented on docs/source/guide/index.rst:
```diff
@@ -380,6 +380,342 @@ Using another computer with access to ``repository-service-tuf-api``
+    │ --help              -h        Show this message and exit.                                                                                          │
+    ╰────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯
+
+You can start the metadata update ceremony by providing value for the required
+``--current-root-ur`` argument when you call the ceremony or you will be asked
+about this:
```
(if you want to leave it the comment, you should fix typos, and I also have a minor rewording suggestion)

kairoaraujo commented on docs/source/guide/index.rst:
```diff
@@ -380,6 +380,342 @@ Using another computer with access to ``repository-service-tuf-api``
+
+    ┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
+    ┃                                      STEP 1: Authorization                                       ┃
+    ┗━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┛
```
Isn't it better to start by loading its signing key? Otherwise, the user experience can have a bad experience answering/changing everything, and in the end, they notice that they need the key or the key is ready to be used.

lukpueh replied:
I'd support a way to store changed metadata, even if it isn't fully signed yet.

MVrachev replied:
This is worth discussing for the https://github.com/repository-service-tuf/repository-service-tuf-cli/issues/294 feature.

lukpueh reviewed (commented):
Amazing work, Martin! Here's another pure UX review based on index.rst. I think what you have works well and I'm happy to move on to an actual code review.

I already mentioned that key loading of previous root signing keys could happen in the end (at signing time). Same is true for root singing keys and online signing key. See inline comments for why I think adding public keys and signing should be separated.

Below is how I'd picture the workflow. No need to streamline too much at this point. But maybe you want to adopt some ideas? 

1. Enter path to previous root (bonus: default to downloading latest root with ngclient)
1. Enter bump root expiration 
1. Enter threshold
1. Until break (and number of keys >= threshold):
    1. Enter (1) remove root key (2) add root key (3) done
    1. If (1): Enter keyid/name to remove key
    1. If (2): Enter path to new public key + name
    1. If (3): Break
 1. Enter
 1. Until number of signatures from previous keys >= previous threshold and number of signatures from new keys >= new threshold:
    1. Enter (1) sign (2) just save
    1. If (1): Enter path to signing key and password
    1. If (2): Enter path to output file and quit workflow
1. Enter should upload
1. Enter should save
    1. Enter path to output file



Additionally I suggest two subcommands:
- `rstuf admin metadata upload <path/to/metadata>` to upload metadata that hasn't been uploaded
- `rstuf admin metadata sign <path/to/metadata` to sign metadata that isn't fully signed

kairoaraujo commented on docs/source/guide/index.rst:
```diff
@@ -380,6 +380,342 @@ Using another computer with access to ``repository-service-tuf-api``
+
+    Choose root key type [ed25519/ecdsa/rsa] (ed25519):
+    Enter the root`s private key path: rstuf/cli/tests/files/key_storage/JanisJoplin.key
+    Enter the root`s private key password:
```
> Amazing work, Martin! Here's another pure UX review based on index.rst. I think what you have works well and I'm happy to move on to an actual code review.
> 

Hi @lukpueh, thank you for the valuable comments!

> I already mentioned that key loading of previous root signing keys could happen in the end (at signing time). Same is true for root singing keys and online signing key. See inline comments for why I think adding public keys and signing should be separated.

As we still don't have the signing feature in place at the moment, we have the keys to make sure who is starting the Metadata Update is a key holder. Also, it keeps aligned with the current ceremony flow.
The good thing is our thinking is aligned with your suggestions.

As you can see in the issue [Task: Add distributed async support in the ceremony, metadata update, and add feature metadata sign](https://github.com/repository-service-tuf/repository-service-tuf-cli/issues/294) that is part of the [Feature: support distributed async threshold signing](https://github.com/repository-service-tuf/repository-service-tuf/issues/327) we no longer ask the users to load`*` the keys but providing the public keys, keyid and signing separated. The online key definitly doesn't require to be loaded.

`*` Only for who starts the `Ceremony`, `Metadata Update` should load its key, which make they an "authorized" or "key owner" to start the process.

Also, we include the `rstuf admin metadata sign` in the above task.

I think adding a `rstuf admin metadata upload` could be also a new CLI `rstuf admin metadata` feature, to work with small tasks.

kairoaraujo commented on docs/source/guide/index.rst:
```diff
@@ -380,6 +380,342 @@ Using another computer with access to ``repository-service-tuf-api``
+    │ --run-ceremony                When '--upload' is set this flag can be used to run the ceremony and the result will be uploaded.                    │
+    │ --save              -s        Save a copy of the metadata locally. This option saves the JSON metadata update payload file in the in the current   │
+    │                               directory.                                                                                                           │
+    │ --help              -h        Show this message and exit.                                                                                          │
```
First of all, I agree entirely with the suggestion.

However, in this PR, we are not adding or removing those parameters for the `rstuf admin ceremony`. I think it is an update for the documentation only. We should file an issue and add it as an improvement for the `rstuf admin ceremony` parameters, and we can decide if we want these changes on MWV (beta) or later.

The changes for the `rstuf admin metadata upload` could also be a new issue, a small new feature, to avoid extending (make even bigger this PR). If it is a decision to add to this PR, I'm ok as well, but doing it in a specific PR could be faster.

MVrachev replied:
As @kairoaraujo has suggested, let's keep the current state as is given that it works and is tested.
I filed an issue about that: https://github.com/repository-service-tuf/repository-service-tuf-cli/issues/298 and we can think about how to solve this problem for both ceremonies.

MVrachev commented on docs/source/guide/index.rst:
```diff
@@ -380,6 +380,342 @@ Using another computer with access to ``repository-service-tuf-api``
+    Do you want to change the online key? [y/n]: y
+
+    Choose root key type [ed25519/ecdsa/rsa] (ed25519): rsa
+    Enter the root`s private key path: rstuf/cli/tests/files/key_storage/online-rsa.key
```
I added a new issue https://github.com/repository-service-tuf/repository-service-tuf-cli/issues/299  where we can fix this problem for both `bootstrap` and `metadata update` ceremonies.

MVrachev commented on docs/source/guide/index.rst:
```diff
@@ -380,6 +380,342 @@ Using another computer with access to ``repository-service-tuf-api``
+    │ Threshold: 1                 │   800dfb5a1982b82b7893e58035e19f414f553fc08cbb1130cfbae302a7b7fee5 │ Steven's Key │ ed25519  │ Offline │ 7098f769f6ab8502b50f3b58686b8a042d5d3bb75d8b3a48a2fcbc15a0223501   │
+    │ Root Expiration: 2024-May-28 │   1cebe343e35f0213f6136758e6c3a8f8e1f9eeb7e47a07d5cb336462ed31dcb7 │ Kairo's Key  │ ed25519  │ Offline │ ad1709b3cb419b99c5cd7427d6411522e5a93aec6767453e91af921a73d22a3c   │
+    │                              │                                                                    ╵              ╵          ╵         ╵                                                                    │
+    └──────────────────────────────┴─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
```
When you remove I key I immediately show you how the root keys look.
Same if you add a key.
It's hard to believe that you deleted the wrong key if you have used key names (which I think many people will use) further it's even harder to make a mistake adding a key when we ask you about this new key path, password and type.
If for whatever reason you see there is a problem I give the opportunity to modify the root keys yet again.

Maybe the most inconvenient is that if you remove a key it could be hard to add it again (if you did a mistake) if you don't have access to that key.
In this case you can always abort the operation, but if we decide it's still a risk we can ask the user for confirmation.
Still, if the user deletes a key that doesn't have a name, then even if we show him the key id of the key for him it will be nonsensical and that confirmation doesn't bring a lot of value.

Finally, it's hard to make a mistake and delete a key with a name as we ask the user to write down the name of the key for deletion.

@kairoaraujo any opinions?

kairoaraujo replied:
I think what you @MVrachev sounds good for this version.
If we still think we can improve the UX we could file an issue.

We know that the CLI there is some limitations as well.
I think some interactions will also change with the possibility of distributed asynchronous signing.

kairoaraujo replied:
@lukpueh, please do it. ❤️

MVrachev commented:
These days I will push a couple of new commits addressing your comments and adding both unit and full integration tests.

kairoaraujo commented on tests/unit/cli/admin/test_md_update.py:
Could you please rename the file to `test_metadata_update.py` to follow the pattern for the test, which is `tests/{test_type: unit|function|integration}/{file path}/test_{filename}.py`

The same for the test functions `def test_{function_name}_{scenario|case}`

lukpueh reviewed (commented):
Here's a review of `repository_service_tuf/helpers/api_client.py`

lukpueh commented on repository_service_tuf/cli/admin/metadata.py:
```diff
@@ -0,0 +1,626 @@
+            "Do you want to extend the [cyan]root's expiration[/]?"
+        )
+        if change is False:
+            if current_root.expiration < (datetime.now() + timedelta(days=1)):
```
It's not very user-friendly to first give the user a choice and then check if they even have a choice. If they have to change it anyway, if it's expired, you can ask them to change right away.

lukpueh reviewed (changes requested):
Thanks for all the hard work, @MVrachev!

I went through all your code changes and additions, except tests, and I think this should work fine. 

I didn't comment on UX and flow, because we already discussed this in my previous review and decided to not change anything at this point.

Most comments are small linter-style simplification suggestions, like using implicit boolness, or removing unnecessary else blocks after raise/return/break/continue etc. I pointed out most of the places, where this could be changed, but not all. Maybe you can give it another pass and see if you can flatten the control flow a little bit? IIRC, pylint should actually flag some of those too.

I also left a few more architectural comments. Especially the conflation of public keys and signing keys in RSTUFKey does not seem ideal. IMO it would make the code a lot easier to reason about, if you e.g. used RSTUFKey only for signing keys.

lukpueh commented on repository_service_tuf/helpers/tuf.py:
```diff
@@ -57,6 +78,181 @@ def to_dict(self):
+        self._newly_added_signing_keys = {}
+
+    @classmethod
+    def _get_name(cls, key: Union[Key, RSTUFKey]) -> str:
```
Before you change this, please read my comment about RSTUFKey in the review summary.

lukpueh commented on repository_service_tuf/helpers/tuf.py:
```diff
@@ -57,6 +78,181 @@ def to_dict(self):
+                name = key.unrecognized_fields["name"]
+
+        elif isinstance(key, RSTUFKey):
+            if key.name is None or key.name == "":
```
Before you change this, please read my comment about RSTUFKey in the review summary.

MVrachev replied:
I change it to a different method and changed how I check about this as a whole.
If you still don't like it leave a comment here.
Otherwise, I still think it's better to make sure there is a name here as well.

MVrachev commented on repository_service_tuf/helpers/api_client.py:
```diff
@@ -201,3 +206,82 @@ def publish_targets(settings: LazySettings) -> str:
+    ):
+        raise click.ClickException(response.text)
+
+    else:
```
We have not enabled `pylint` linter.

MVrachev commented on repository_service_tuf/helpers/api_client.py:
```diff
@@ -201,3 +206,82 @@ def publish_targets(settings: LazySettings) -> str:
+def send_payload(
+    settings: LazySettings,
+    url: str,
+    method: Methods,
```
I agree, will change it.

MVrachev commented on repository_service_tuf/helpers/tuf.py:
```diff
@@ -57,6 +78,181 @@ def to_dict(self):
+        name = cls._get_name(tuf_online_key)
+        online_key = RSTUFKey(online_key_dict, name=name)
+
+        return cls(root_md, root_keys, online_key)
```
The one place where `root_keys` is useful when creating a keys table when calling the `_create_keys_table` function here when printing everything related to the root role: https://github.com/MVrachev/repository-service-tuf-cli/blob/540f41f2adc0e6e5d5438f17fd161a7b9a4ecd99/repository_service_tuf/cli/admin/metadata.py#L162
and here:
https://github.com/MVrachev/repository-service-tuf-cli/blob/540f41f2adc0e6e5d5438f17fd161a7b9a4ecd99/repository_service_tuf/cli/admin/metadata.py#L266
when printing all of the keys (including non-singing keys) in order to decide which of the keys you want to remove.

MVrachev replied:
I will try my best to find a workaround and use `RSTUFKey` only for signing keys.

MVrachev commented on repository_service_tuf/helpers/tuf.py:
```diff
@@ -57,6 +78,181 @@ def to_dict(self):
+        for keyid in root_md.signed.roles[Root.type].keyids:
+            tuf_key: Key = root_md.signed.keys[keyid]
+            name = cls._get_name(tuf_key)
+            tuf_key.unrecognized_fields["name"] = name
```
Yes, it is. 
We want to have `names` for all root keys even if the user doesn't provide one.
This helps later when removing a key as the user gives us a name to the key he wants to remove and even if the key doesn't have a name initially we want to give a name to the key, so the user can easily write down when he wants to remove it.
Additionally, when we have to remove the key later we can store the keys in a dict where the keys are the name of the corresponding root keys.

Finally, this won't invalidate the check if `root` was changed because we copy into `_initial_root_md_obj` as the final step of the initialization after all keys have names added to them.

MVrachev commented on repository_service_tuf/cli/admin/metadata.py:
```diff
@@ -0,0 +1,626 @@
+                    "Do you want to abort the root metadata update",
+                )
+                if abort:
+                    raise click.ClickException(
```
Okay, yes.
I will just not give the user the choice and directly inform him that he needs to add new keys and ask him for the relevant information.
If he wants to quit he can always do that with `cntrl + c`.

MVrachev replied:
I decided to be consistent and if I need any update/information I will tell the user that it's required and just ask for it, not asking the user specifically if he wants to give it to me and then fail with an exception.

MVrachev commented:
@lukpueh and @kairoaraujo: I followed Lukas suggestions to:
- simplify the workflow at multiple places
- fixed unti tests mockind leakage causing problems
- stopped using RSTUFKey as a container for non-signing keys. RSTUFKey is used only for loaded or added root keys that will be used for signing
- fixed typos, boolean checks 
- fixed linting

lukpueh reviewed (commented):
Here's a quick review of `RootInfo`

lukpueh commented on repository_service_tuf/cli/admin/metadata.py:
```diff
@@ -0,0 +1,605 @@
+        )
+        if agree:
+            current_root.expiration = new_expiry
+            return
```
nit: I don't think you need two loops here. what about something like...

```python
while True:

    if root.is_expired:  # this may already exist in Metadata API, if not you could add it to RootInfo
        print("root expired. please change!")
        change = True

    else:
        change = prompt(f"root expires on {root.expiration}. change? y/n") # expiration getter could do the formatting

    if not change:
        break

    root.expiration = prompt("enter days from now") # setter could do the math
```

lukpueh replied:
This is roughly the same flow, which you have in `_keys_additions`, and which I think is pretty good. 👍

MVrachev replied:
I think I found a way to simplify it to one cycle.

lukpueh commented on repository_service_tuf/cli/admin/metadata.py:
```diff
@@ -0,0 +1,605 @@
+    input: int = 0
+    while True:
+        input = prompt.IntPrompt.ask(msg, default=default, show_default=True)
+        if input < 1:
+            console.print(f"{input_name} must be at least 1")
+        else:
+            break
+
+    return input
```
nit: this could be
```suggestion
        if input >= 1:
            return input

        console.print(f"{input_name} must be at least 1")

```

lukpueh commented on repository_service_tuf/cli/admin/metadata.py:
```diff
@@ -0,0 +1,605 @@
+                "\n", f":cross_mark: [red]Failed[/]: key {name} is not in root"
+            )
+            continue
+        else:
```
else not needed

lukpueh replied:
The PR still has plenty of these. 🤷 Probably does not matter.

lukpueh commented on repository_service_tuf/cli/admin/metadata.py:
```diff
@@ -0,0 +1,605 @@
+            continue
+        else:
+            console.print(f"Key with name/tag [yellow]{name}[/] removed\n")
+            if len(current_root.keys) < 1:
```
IMO this check should be before you ask the user, if they want to remove a key. Otherwise, you could again have a situation, where you give a choice (want to remove y/n?) and then inform the user that they don't actually have a choice (remaining keys).

Maybe this is not a problem, because the loop cannot be entered with metadata that has no keys. But it seems easy to change this loop so that it does not require that precondition.

lukpueh commented on repository_service_tuf/cli/admin/metadata.py:
```diff
@@ -0,0 +1,605 @@
+
+
+def _keys_removal(current_root: RootInfo):
+    """Asking the user if he wants to remove any of the root keys"""
```
```suggestion
    """Asking the user if they want to remove any of the root keys"""
```

lukpueh commented on repository_service_tuf/cli/admin/metadata.py:
```diff
@@ -0,0 +1,605 @@
+                raise click.ClickException(
+                    f"Failed authorization. Required: {threshold}, loaded "
+                    f"{len(loaded)} keys"
+                )
```
As I said before, the user shouldn't be given a choice, where one of the options leads to a failure.

Maybe this is only a wording problem. You could tell them to either try again, or exit, indicating the consequences, i.e. no metadata change or upload. Then, when you do exit, you don't have to say that something failed.

But as I also said before, it seems inconsistent to give the option to exit in a few places but require `ctrl+c` in others.

MVrachev replied:
You have a point. Will fix that.

lukpueh reviewed (changes requested):
This is looking a lot better. Here are a few more comments about the CLI.

MVrachev commented on repository_service_tuf/helpers/api_client.py:
```diff
@@ -201,3 +205,73 @@ def publish_targets(settings: LazySettings) -> str:
+    else:
+        role_md = Metadata.from_file(file_uri)
+
+    return role_md
```
Oh, it seems to be a mistake I made.

MVrachev commented on repository_service_tuf/helpers/tuf.py:
```diff
@@ -57,6 +85,151 @@ def to_dict(self):
+        # Make sure every root key will have a name
+        for keyid in self._root_md.signed.roles["root"].keyids:
+            key = self._root_md.signed.keys[keyid]
+            key.unrecognized_fields["name"] = self._get_key_name(key)
```
Okay, I have to agree here.
You are right, it's not correct to change root metadata even if we see a benefit from it.
I can change our code to use key names for all keys internally, but not save this in the metadata.

MVrachev commented:
I addressed all comments except this one: https://github.com/repository-service-tuf/repository-service-tuf-cli/pull/284#discussion_r1234195608
as I didn't want to rush it.
I realized that in order to do it well, we need to warn the user of how many currently loaded signing keys be removed.

MVrachev commented:
I will be PTO until the end of the week, but I will appreciate it if you spend time reviewing my latest changes @kairoaraujo and @lukpueh.

lukpueh commented on repository_service_tuf/helpers/tuf.py:
```diff
@@ -57,6 +85,151 @@ def to_dict(self):
+            self._root_md.sign(SSlibSigner(key.key), append=True)
+
+        console.print("\nVerifying the new payload...")
+        self._root_md.verify_delegate(Root.type, self._root_md)
```
> I will need to find a solution to this by storing the original threshold and making sure there is at least a threshold of current keys loaded at the time of new root key generation.

You already have the solution in your RootInfo:
- `_initial_root_md_obj` has the previous threshold and keys
- `_root_md` has the new threshold and keys

Check-out how @kairoaraujo is [doing this in `_trusted_root_update`](https://github.com/repository-service-tuf/repository-service-tuf-worker/blob/33d74e9d4e2333d19b59471fa350ba899aedd855/repository_service_tuf_worker/repository.py#L992-L996):

lukpueh replied:
minor naming nit: `_initial_root_md_obj` is not ideal. In python-tuf we call the duo "trusted_root" and "new_root". IMO "old" and "new" is also fine. Calling one of them "current" is fine, but a bit ambiguous, because it would fit both. 😄

MVrachev replied:
> > I will need to find a solution to this by storing the original threshold and making sure there is at least a threshold of current keys loaded at the time of new root key generation.
> 
> You already have the solution in your RootInfo:
> 
> * `_initial_root_md_obj` has the previous threshold and keys
> * `_root_md` has the new threshold and keys
> 
> Check-out how @kairoaraujo is [doing this in `_trusted_root_update`](https://github.com/repository-service-tuf/repository-service-tuf-worker/blob/33d74e9d4e2333d19b59471fa350ba899aedd855/repository_service_tuf_worker/repository.py#L992-L996):

I don't think that's correct. I even think there is a need to fix his implementation as well.
The reason is that the specification clearly says:
`if there is a 1.root.json that has threshold 2 and a 2.root.json that has threshold 3, 2.root.json MUST be signed by at least 2 keys defined in 1.root.json and at least 3 keys defined in 2.root.json` meaning we need to make sure that:
1. there are at least **_current_root_threshold_** a number of keys left in the new root metadata that are also defined in the current (trusted) root metadata.
2. there are at least **_new_root_threshold_** of keys in the new root metadata

What Kairo validates is that `current_root` has a `_current_root_threshold_` of trusted keys and `new_root_threshold` has a `_new_root_threshold_` of keys.
This doesn't make the connection between `_current_root_threshold_`  and `new_root`.

lukpueh replied:
I'm pretty sure that Kairo's code does exactly what it should do. It's the same [what ngclient does, when updating root](https://github.com/theupdateframework/python-tuf/blob/e54ee24684e1898cae159484d915ab6a695206be/tuf/ngclient/_internal/trusted_metadata_set.py#L163-L173). You can take a look at the `Metadata.verify_delegate` implementation, if you want to make sure.

MVrachev replied:
Ohhh... how could have missed it.
I totally missed that in this line:
[`current_root.verify_delegate(Root.type, new_root)`](https://github.com/repository-service-tuf/repository-service-tuf-worker/blob/33d74e9d4e2333d19b59471fa350ba899aedd855/repository_service_tuf_worker/repository.py#L993)
we are passing the NEW root and thus verifying that the keys in the current root are actually signing the new root and that their amount is enough.
That does make sense.
Thanks for your patience @lukpueh!

MVrachev replied:
On that note:

> minor naming nit: _initial_root_md_obj is not ideal. In python-tuf we call the duo "trusted_root" and "new_root". IMO "old" and "new" is also fine. Calling one of them "current" is fine, but a bit ambiguous, because it would fit both. 😄

I think I will use the opportunity to start naming the `current root` to `trusted root` and  `_initial_root_md_obj` will become `trusted_root` internally.

lukpueh replied:
> I think I will use the opportunity to start naming the `current root` to `trusted root` and `_initial_root_md_obj` will become `trusted_root` internally.

Wait, you want to call one `trusted root` and the other one `trusted_root`? I don't think that's a good idea.

MVrachev replied:
See my latest changes below.

lukpueh commented on repository_service_tuf/cli/admin/metadata.py:
```diff
@@ -0,0 +1,596 @@
+            new_expiry = datetime.now() + timedelta(days=bump)
+            new_exp_str = new_expiry.strftime("%Y-%b-%d")
+            agree = prompt.Confirm.ask(
+                f"New root expiration: [cyan]{new_exp_str}[/]. Do you agree?"
```
Probably a matter of taste, but I think it's nicer to re-use the prompts from above as I suggested. The flow is a bit awkward if the user does not agree here. It will look like.
```
...
New root expiration: <DATE>. Do you agree?
Current root expiration: <DATE>
Do you want to extend?
...
```

lukpueh replied:
> Probably a matter of taste, but I think it's nicer to re-use the prompts from above[ as I suggested](https://github.com/repository-service-tuf/repository-service-tuf-cli/pull/284#discussion_r1233898921). The flow is a bit awkward if the user does not agree here. It will look like.
> 
> ```
> ...
> New root expiration: <DATE>. Do you agree?
> Current root expiration: <DATE>
> Do you want to extend?
> ...
> ```

MVrachev replied:
I am not sure I understand your idea here, but can it wait for another pr?
Feel free to suggest a pr on top of mine.

MVrachev commented:
I made one important change by improving the signature verification and one smaller one by renaming some of the attributes.
I made a new issue for improving the signature verification process: #306.

lukpueh reviewed (commented):
I find it extremely difficult to keep track of the various places, where you store and read keys from. And I fear that the blurry separation between public keys in old and new root metadata, and signing keys is prone to bugs (see inline comment).

I recommend that you use the public key stores as single sources of truth for what keys are in old and in new root metadata.

Here's how I would present the keys to the user. It works for trusted root and new root alike.
```python
def _create_keys_table(root: Metadata, signing_keys: dict[RSTUFKey]):

   for key in root.keys:
      can_sign = False
      if key.keyid in signing_keys:
         can_sign = True

      # print table row with key infos (id, name/tag, can_sign, etc..)
      # optionally print online/offline info by looking up keyid in root.roles
```

And here is how I would track key changes:
- When loading a signing key for trusted root you only add it to `signing_keys`
- When adding a key for new root, you add it in _new_root.keys (or replace, if already there) and in signing_keys (or replace, if already there).
- When removing a key for new root, you remove it from _new_root.keys only. The related private key may still be used for trusted root. (you could remove it conditionally, if not needed for trusted root, but I don't see a problem, if you just keep it in memory)

MVrachev commented on repository_service_tuf/cli/admin/metadata.py:
```diff
@@ -0,0 +1,596 @@
+        signing_list = [
+            key.to_dict() for key in current_root.signing_keys.values()
+        ]
+        keys_table = _create_keys_table(signing_list, current_root, False)
```
I changed my implementation, but I still use the signing_keys dict.
Now, I make sure that no keys deleted by the user will be shown to the user, but when we calculate how many keys needed to be added I utilize all keys no matter if marked for deletion by the user.

For more information about my motives read my commit message on [this commit](https://github.com/repository-service-tuf/repository-service-tuf-cli/pull/284/commits/5b50bfc326bf657c1042d0c67a7507864a8480d0).

MVrachev commented:
I added the last commit to resolve the CI failures when running `tox -e requirements`, but I think we put `securesystemslib[cryptoy]` and `dynaconf[ini]` with `extras` to make securesystemslib use `crypto` and `dynaconf` use `ini`.
Is that okay that I removed the extras from the requirements files or should I fix the CI error in another way? 
The error that I got before were those:
```
< dynaconf[ini]==3.1.12
---
> dynaconf==3.1.12
84c84
< securesystemslib[crypto]==0.28.0
---
> securesystemslib==0.28.0
```

kairoaraujo commented:
> I added the last commit to resolve the CI failures when running `tox -e requirements`, but I think we put `securesystemslib[cryptoy]` and `dynaconf[ini]` with `extras` to make securesystemslib use `crypto` and `dynaconf` use `ini`. Is that okay that I removed the extras from the requirements files or should I fix the CI error in another way? The error 

Update your `pipenv` package.
There were some changes in how to generate the requirements.txt

MVrachev commented:
> > I added the last commit to resolve the CI failures when running `tox -e requirements`, but I think we put `securesystemslib[cryptoy]` and `dynaconf[ini]` with `extras` to make securesystemslib use `crypto` and `dynaconf` use `ini`. Is that okay that I removed the extras from the requirements files or should I fix the CI error in another way? The error
> 
> Update your `pipenv` package. There were some changes in how to generate the requirements.txt
> 
> So, you can run `make requirements` and it will fix.

I already did that in https://github.com/repository-service-tuf/repository-service-tuf-cli/pull/284/commits/8eb9a6176fa96e745f8c9ea00e6629a6c58afa21, but this didn't help.
After the requirements were updated an error appeared in the CI: https://github.com/repository-service-tuf/repository-service-tuf-cli/actions/runs/5454841078/jobs/9925493014.

What is strange is that if I fix manually what is needed as I did in my commit https://github.com/repository-service-tuf/repository-service-tuf-cli/pull/284/commits/c594160f0e623e86a4cae94f6c69b83677d8a3e2 which fixed the CI, but then run `make requirements`, then this change will revert to the old state which is using:
`dynaconf[ini]==3.1.12` and `securesystemslib[crypto]==0.28.0`.

MVrachev commented:
I will rebase when there is approval on my changes.
Otherwise, the commit history will be lost and it won't be clear which of the commits are new and when they were added.

MVrachev commented:
> Here's a hands-on review, based on running `rstuf admin metadata update` with [root metadata](https://github.com/theupdateframework/python-tuf/blob/5dddfe28aad8fa6261be9324f396211f350c83f3/tests/repository_data/repository/metadata/1.root.json) and [key files](https://github.com/theupdateframework/python-tuf/tree/5dddfe28aad8fa6261be9324f396211f350c83f3/tests/repository_data/keystore) from the python-tuf test repo and trying different inputs.
> 
> First of all, the cli dialogues are really pretty and AFAICT from the UI the tool seems to mostly work, but I stumbled across two bugs:
> 
> One bug was that the tool showed me a wrong number of keys needed to reach the threshold, after doing the following:
> 
> 1. load root with 1 key and threshold of 1,
> 2. "authorise" the key
> 3. change threshold to 2
> 4. remove key from new root
> 5. tool says "You must add 1 more key(s)" but should say: "You must add 2 more key(s)"

I don't think this is a bug.
This happens because I don't actually need you to add two keys. I need you to add one more.
The key you authorized yourself with even if removed will be used by RSTUF for signing and we need one additional key.

> The other bug was an exception I got when I tried to change the online key: `ValueError: Key with id 8a1c4a3ac2d515dec982ba9910c5fd79b91ae57f625b9cff25d06bf0a61c1758 is not used by snapshot`
> 
> I did not debug this, but from a quick look at the line where the error got raised, it seems like you are using a method that is intended for root key removal to remove the online key?
> 
> Here's a [terminal dump](https://github.com/repository-service-tuf/repository-service-tuf-cli/files/11959245/20230705_review-lp_rstuf-admin-metadata-update.log), which includes the faulty prompt (line 138) and the full stack trace (line 536)

This is worth investigating.

kairoaraujo commented:
I spotted this issue showing duplicated *Failed* word
```console
Choose online key type [ed25519/ecdsa/rsa] (ed25519): 
Enter the online`s private key path: fdfd
Enter the online`s private key password: 
❌ Failed: ❌ Failed: Can't open fdfd
```

kairoaraujo commented:
I also couldn't reproduce the `ValueError` exception.

MVrachev commented:
I rebased on top of main. Ready for full review @kairoaraujo.

kairoaraujo reviewed (approved):
LGTM

kairoaraujo commented:
I ran the functional tests, and the consistency is great!

```shell
------------------------------------------------------------------------------------------------- generated json file: /Users/kdearaujo/devel/RSTUF/repository-service-tuf/test-report.json --------------------------------------------------------------------------------------------------
---------------------------------------------------------------------------------------------- generated html file: file:///Users/kdearaujo/devel/RSTUF/repository-service-tuf/test-report.html ----------------------------------------------------------------------------------------------
===================================================================================================================================== slowest durations ======================================================================================================================================
195.63s call     devel/RSTUF/repository-service-tuf/tests/functional/targets/test_performance.py::test_api_requester_multiple_request_and_targets[50-50-300]
61.47s call     devel/RSTUF/repository-service-tuf/tests/functional/targets/test_performance.py::test_api_requester_multiple_request_and_targets[100-2-180]
26.75s call     devel/RSTUF/repository-service-tuf/tests/functional/targets/test_performance.py::test_api_requester_multiple_request_and_targets[2-100-60]
17.48s call     devel/RSTUF/repository-service-tuf/tests/functional/targets/test_performance.py::test_api_requester_multiple_request_and_targets[5-10-30]
6.58s call     devel/RSTUF/repository-service-tuf/tests/functional/targets/test_remove_targets.py::test_removing_a_target_using_rstuf_api[["file1.tar.gz", "a/file2.tar.gz", "c/d/file3.tar.gz"]]
6.47s call     devel/RSTUF/repository-service-tuf/tests/functional/targets/test_remove_targets.py::test_removing_a_target_using_rstuf_api[["file1.tar.gz", "a/file2.tar.gz"]]
6.44s call     devel/RSTUF/repository-service-tuf/tests/functional/targets/test_remove_targets.py::test_removing_targets_that_does_exist_and_ignoring_the_rest[["file1.tar.gz", "a/file2.tar.gz"]-["foo", "bar"]]
6.35s call     devel/RSTUF/repository-service-tuf/tests/functional/targets/test_remove_targets.py::test_removing_targets_that_does_exist_and_ignoring_the_rest[["file1.tar.gz"]-["foo"]]
6.29s call     devel/RSTUF/repository-service-tuf/tests/functional/targets/test_performance.py::test_api_requester_multiple_request_and_targets[2-2-10]
6.29s call     devel/RSTUF/repository-service-tuf/tests/functional/targets/test_remove_targets.py::test_removing_a_target_using_rstuf_api[["file1.tar.gz"]]
3.15s call     devel/RSTUF/repository-service-tuf/tests/functional/targets/test_add_targets.py::test_adding_a_target_using_rstuf_api[630-"716f6e863f744b9ac22c97ec2f677b76ea5f5908bc5bc61510bfc4751384ea7a"-{"key": "value"}-"file1.tar.gz"]
3.10s call     devel/RSTUF/repository-service-tuf/tests/functional/targets/test_add_targets.py::test_adding_a_target_using_rstuf_api[2024-"93ea575cb5d8a053eaa0ac8fa3b40d7e05a33cc853eaa0ac8fa3b46f6e863f74"-None-"a/file2.tar.gz"]
3.09s call     devel/RSTUF/repository-service-tuf/tests/functional/targets/test_add_targets.py::test_adding_a_target_using_rstuf_api[532-"d9f34f8cd5cb3b3eb79b3e4b5dae3a16df499a70d8a053eaa0ac8fa5f5908bc5"-{"key": "value"}-"a/b/file3.tar.gz"]
1.55s call     devel/RSTUF/repository-service-tuf/tests/functional/bootstrap/test_bootstrap.py::test_bootstrap_using_rstuf_command_line_interface_cli
1.43s call     devel/RSTUF/repository-service-tuf/tests/functional/bootstrap/test_bootstrap.py::test_bootstrap_using_rstuf_cli_with_invalid_payload
0.15s call     devel/RSTUF/repository-service-tuf/tests/functional/targets/test_remove_targets.py::test_removing_targets_that_does_exist_and_ignoring_the_rest[[]-["foo", "bar"]]
=============================================================================================================================== 16 passed in 352.78s (0:05:52) ===============================================================================================================================
```

kairoaraujo commented:
I'm merging this feature.
If someone has any other comments we can address them as issues as new interactions.
